File: Programmers/level-1/키패드누르기.py
def solution(numbers, hand):
    numArray = [[1,2,3],[4,5,6],[7,8,9],[-1,0,-1]] # 지도
    left = [3, 0]  # [x,y]
    right = [3, 2]  # [x,y]
    
    answer = ''
    for findNum in numbers: # 누를 번호 반복하기
        # 지도에서 누를 번호의 위치 찾기
        findCoor = [[i,j] for i in range(4) for j in range(3) if numArray[i][j] == findNum][0]
        
        n = numArray[findCoor[0]][findCoor[1]]
        if n == 1 or n ==4 or n ==7:
            left = findCoor
            answer += "L"
        elif n == 3 or n == 6 or n == 9:
            right = findCoor
            answer += "R"
        else:
            
            findX, findY = findCoor[0], findCoor[1]

            # 왼손, 오른손 현재 위치와 각각 누를 위치와 차이 구하기
            leftDif = abs(left[0]-findX) + abs(left[1]-findY)
            rightDif = abs(right[0]-findX) + abs(right[1]-findY)
            
            if leftDif>rightDif:
                right = findCoor
                answer+="R"
                
            elif rightDif>leftDif:
                left = findCoor
                answer+="L"
            else:
                if hand == "left":
                    left = findCoor
                    answer+="L"
                else:
                    right = findCoor
                    answer+="R"
                
    return answer

File: java/프로그래머스/level-1/키패드누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder answer = new StringBuilder();
        int left = 10, right = 12;
        
        for (int num : numbers){
            if (num==0) num =11;
            
            switch (num %3){
                    
                    case 0->{
                        answer.append("R");
                        right = num;
                        
                    }
                    
                    case 1->{
                        answer.append("L");
                        left = num;
                        
                    }
                    
                    case 2->{ // L or R
                        int minusL = Math.abs(num-left); // 현재 왼손 위치와 숫자의 차이
                        int minusR = Math.abs(num-right); //현재 오른손 위치와 숫자의 차이
                        int distanceL = (minusL %3) +(minusL/3);
                        int distanceR = (minusR %3) +(minusR/3);
                        
                        if (distanceL < distanceR){
                            answer.append("L");
                            left = num;
                        }else if (distanceL > distanceR){
                            answer.append("R");
                            right = num;
                        }else{
                            answer.append(hand.equals("right") ?"R":"L");
                            if (hand.equals("right")) right = num;
                            else left = num;
                        }
                    }
            }
        }
        return answer.toString();
    }
}